internal/statuspage/component_api_mock.go

```diff
@@ -0,0 +1,151 @@
+package statuspage
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/broadinstitute/revere/internal/configuration"
+	"github.com/jarcoal/httpmock"
+	"math/rand"
+	"net/http"
+	"strconv"
+)
+
+//goland:noinspection SpellCheckingInspection
+var letterRunes = []rune("abcdefghijklmnopqrstuvwxyz")
+
+// randStringRunes makes a random alphabetic string of the given length
+func randStringRunes(length int) string {
+	b := make([]rune, length)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(b)
+}
+
+// componentFactory helps make a unique component based on name alone.
+// Creates an ID for the component to simplify testing, so that tests don't
+// need to preface tests with information-gathering GETs to the mock.
+// API functions themselves will strip out the ID before the component goes to
+// the server.
+func componentFactory(name string) *Component {
+	return &Component{
+		ID:     randStringRunes(8),
+		Name:   name,
+		Status: "operational",
+	}
+}
+
+// validatePageID returns a 404 response if the pageID wasn't present as the first regex of the request URL,
+// and nil otherwise
+func validatePageID(pageID string, request *http.Request) *http.Response {
+	reqPageID := httpmock.MustGetSubmatch(request, 1)
+	if reqPageID != pageID {
+		return httpmock.NewStringResponse(
+			404,
+			fmt.Sprintf("Page ID %s was not correct (should be %s)", reqPageID, pageID))
+	}
+	return nil
+}
+
+// validateComponentID returns a 404 response if a component ID wasn't present as the second regex of the request URL,
+// and nil otherwise
+func validateComponentID(components map[string]Component, request *http.Request) *http.Response {
+	reqComponentID := httpmock.MustGetSubmatch(request, 2)
+	for id := range components {
+		if id == reqComponentID {
+			return nil
+		}
+	}
+	return httpmock.NewStringResponse(
+		404,
+		fmt.Sprintf("Component ID %s was not found in components map", reqComponentID))
+}
+
+// ConfigureComponentMock mimics the behavior of Statuspage's component API via the given backing map.
+// Any components given in the initial map or created via the mock will have their page ID properly set.
+// Created component IDs are incremented based on component map size and the number of deleted components.
+// The caller is responsible for activating/deactivating/resetting httpmock.
+func ConfigureComponentMock(config *configuration.Config, components map[string]Component) {
```

Comment:
Having a single function this long triggers my spidey senses. Is it feasible to break this up a bit?

Reply:
I don't know. It is long because it is registering the handlers for the four request types, and I'm not sure I can break it up because all four need to reference some shared data store. The handler function type doesn't allow that sort of data exchange, and I'm wary of using variables outside of a function because this is scoped within the httpmock "activate" and "deactivateAndReset" sequence.

Frankly this mock is complex enough that we could potentially literally use a fully-blown server library to handle this, and just serve it on localhost during the testing, but I'm worried that that would be more difficult to maintain than this function.